[scroll-animations] update `EffectTiming.duration` to use `CSSNumberish`

graouts opened the pull request:
#### 1e783c0e6d6a20328d171fcbfcc91384ddb0c263
<pre>
[scroll-animations] update `EffectTiming.duration` to use `CSSNumberish`
<a href="https://bugs.webkit.org/show_bug.cgi?id=279728">https://bugs.webkit.org/show_bug.cgi?id=279728</a>
<a href="https://rdar.apple.com/136031070">rdar://136031070</a>

Reviewed by Tim Nguyen.

Add CSSNumericValue as one of the options for `EffectTiming.duration`. Since percentage
values are not allowed as input, we add the `EffectTiming::durationAsDoubleOrString()`
to convert this value for the `double` and `DOMString` case.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CustomEffect.cpp:
(WebCore::CustomEffect::create):
* Source/WebCore/animation/EffectTiming.cpp: Added.
(WebCore::EffectTiming::durationAsDoubleOrString const):
* Source/WebCore/animation/EffectTiming.h:
* Source/WebCore/animation/EffectTiming.idl:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::create):

Canonical link: <a href="https://commits.webkit.org/284623@main">https://commits.webkit.org/284623@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92b902c0f824d9a3b3bec56601ca588f1a843776

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70036 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49439 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22794 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74123 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21202 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72153 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57241 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21045 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/55580 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14059 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73102 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45037 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60409 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/36046 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41697 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17840 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19571 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63621 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18187 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75840 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14261 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17423 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63284 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14301 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60478 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63212 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11227 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4844 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10690 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45244 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46318 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47592 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46059 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/animation/EffectTiming.cpp:
```diff
@@ -0,0 +1,40 @@
+
+std::optional<std::variant<double, String>> EffectTiming::durationAsDoubleOrString() const
+{
+    if (auto* doubleDuration = std::get_if<double>(&duration))
```
Looks like something we'd use `WTF::switchOn()` for in general. I suspect it may be faster too.

graouts replied:
Right on, I will use `WTF::switchOn()` in the commit.

graouts committed:
[scroll-animations] update `EffectTiming.duration` to use `CSSNumberish`
https://bugs.webkit.org/show_bug.cgi?id=279728
rdar://136031070

Reviewed by Tim Nguyen.

Add CSSNumericValue as one of the options for `EffectTiming.duration`. Since percentage
values are not allowed as input, we add the `EffectTiming::durationAsDoubleOrString()`
to convert this value for the `double` and `DOMString` case.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CustomEffect.cpp:
(WebCore::CustomEffect::create):
* Source/WebCore/animation/EffectTiming.cpp: Added.
(WebCore::EffectTiming::durationAsDoubleOrString const):
* Source/WebCore/animation/EffectTiming.h:
* Source/WebCore/animation/EffectTiming.idl:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::create):

Canonical link: https://commits.webkit.org/284623@main
